File: Review-01/Example-3/source/listDriver.cpp
// CS 330 Fall 2016
// Review Example 1: Linked List Review

#include <iostream>
#include <iomanip>
#include <ctime>
#include <cstdlib>
#include <string>

#include "utilities.h"

#include "linkedList.h"

/** @file */

using namespace std;
using namespace utilities;

const string PROGRAM_HEADING[] = {
    "Linked List Review",
    "Thomas J. Kennedy"
}; ///< Program Title 

const int HEADING_LINES = 2; ///< Number of lines in Program Heading

const int MIN = -10; ///< Lower bound for number generation
const int MAX = 10;  ///< Upper bound for number generation

/**
 * Generate a random integer in the range min, max.
 * Default to MIN and MAX
 */
int randomInt(int min=MIN, int max=MAX);

/**
 * Generate a Linked List of random integers
 *
 * @param n number of integers to generate
 */
LinkedList generateList(int n);

/**
 * Linked List Example 2
 */
int main(int argc, char** argv)
{
    int         seed        = 0; // Seed for random number generation
    int         toGenerate  = 0; // Number of nodes to generate    

    // If a seed was passed from the command line,
    // parse it. Otherwise default to ctime
    if( argc >= 2 ) {
        seed = atoi( argv[1] );          
    }
    else {
        seed = time( NULL );
    }

    // If a node  count was passed from the command line,
    // parse it. Otherwise default to 10
    if( argc >= 3 ) {
        toGenerate = atoi( argv[2] );   
    }
    else {
        toGenerate = 10;
    } 

    srand( seed );    

    // Print the program heading
    printProjectHeading( cout, PROGRAM_HEADING, HEADING_LINES );

    // Create a Linked List
    LinkedList randomInts = generateList(toGenerate);
    cout << randomInts;  

    printHorizontalLine( cout, '*', W_WIDTH );

    LinkedList randomCopy = randomInts;

    randomCopy.appendNode(337);

    cout << randomCopy;

    printHorizontalLine( cout, '*', W_WIDTH );
    cout << randomInts; 

    return 0;
}

/**
 *
 */
int randomInt( int min, int max )
{
    // Generate a number in the range 0-1
    double x = (1.0 * rand() / RAND_MAX);

    // Compute scaling factor
    int s = max - min + 1;

    // Multiply x by the scaling factor and shift
    // by adding min
    return ( x * s ) + min; 
}

/**
 *
 */
LinkedList generateList(int n)
{
    LinkedList ll;

    // Generate a Linked List (LL) of 10 Nodes
    for( int i = 0; i < n; i++ ){     
        ll.appendNode( randomInt() );
    }

    return ll;
}

File: Review-01-Python-Linked-Lists/Example-1/list_driver.py
#! /usr/bin/env python3

import sys
import random
import copy

from linkedlist import LinkedList

PROGRAM_HEADING = [
    "Linked List Review",
    "Thomas J. Kennedy"
]


MIN = -10  # Lower bound for number generation
MAX =  10  # Upper bound for number generation


def random_int(min=MIN, max=MAX):
    """
    Generate a random integer in the range min, max.
    Default to MIN and MAX
    """

    return random.randint(min, max)


def generate_list(n):
    """
    Generate a Linked List of random integers

    :param n: number of integers to generate
    """

    ll = LinkedList()

    # Generate a Linked List (LL) of n Nodes
    for i in range(0, n):
        ll.append(random_int())

    return ll


def main():
    """
    Back to the beginning of the semester... with Linked Lists
    """

    to_generate = 0  # Number of nodes to generate

    # If a seed was passed from the command line,
    # parse it. Otherwise default to ctime
    try:
        random.seed(int(sys.argv[1]))

    except IndexError as _err:
        pass

    # If a node count was passed from the command line,
    # parse it. Otherwise default to 10
    try:
        to_generate = int(sys.argv[2])

    except (IndexError, ValueError) as err:
        to_generate = 10

    # Print the program heading
    print("-" * 80)

    for line in PROGRAM_HEADING:
        print(f"{line:^80}")

    print("-" * 80)

    # Create a Linked List
    random_ints = generate_list(to_generate)
    print(random_ints)

    print("*" * 80)

    random_copy = copy.deepcopy(random_ints)

    random_copy.append(337)

    print(random_copy)

    print("*" * 80)
    print(random_ints)


if __name__ == "__main__":
    main()
